### PR TITLE
feat: builds for Go files in a different directory

### PR DESCRIPTION
**Pull request by Harleenijjar**
*Issue #, if available:*
Bug: sam build --use-container doesn't correctly build golang containers #3894
Unable to build go files using --use-container when go files are in different directory than the main file because it only mounts one directory which is the codeUri to the docker container. This implies that only the files in that directory are found when using a container. Works for non container builds because no mounting is required.

*Description of changes:*
Added a method to build go files while using a container when go files are in a different directory than the main files. Method is called when sam build is unable to find the files which will imply that the files are in a different directory. It will run this method and be able to find the main.go files.Therefore, Mounting the parent directory with all the files. This implies the CodeUri must be the parent directory and handler must be the path to the main.go file. This way the build command will be able to find the go files in the parent directory plus follow the path to the main.go file.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by mildaniel — CHANGES_REQUESTED** [review]
Great work! Looks good overall. I left a few nit-picky comments and a question. 

Can we also just add an integration test case which covers the customer's specific use-case to make sure we don't do anything to break that in the future.

**Comment by moelasmar** [comment]
Thanks @Harleenijjar for the contribution. 
I think there is some issue with this approach. In the linked issue, is the customer did not add the go.mod in the same directory as the main.go file, but in a parent directory. 

so, when customer executes `sam build`, SAM CLI will execute `go build` command in the code_uri directory, and `go` command will find the `go.mod` file in the parent directory. But, when customer executes `sam build --use-container`, SAM CLI will mount only the code_uri to the created container, so when SAM CLI executes `go build` command, `go` will not find the `go.mod` file anywhere, as SAM CLI does not mount the parent directories.

so I think using the Handler as a location to run `go build` command will not solve the problem, as actually the Handler is not a location or a directory name, it can be any name the customer choose. The handler is the name of the generated executable file created by `go build` command.

I think the solution is we need the customer to provide a base directory to be mounted to the container, and then we can normally run the `go build` command from the code_uri path. We can suggest adding a new metadata info to be added to the Lambda function to refer to the base directory where the source code is defined, and also the module files.

**Review by moelasmar — DISMISSED** [review]
I left my comment in the conversation

File: aws_lambda_builders/workflows/go_modules/builder.py [hunk 1]
```diff
@@ -69,6 +71,17 @@ def build(self, source_dir_path, output_path):
...
+            LOG.debug("Go files not found. Attempting to build for Go files in a different directory")
+            process, p_out, p_err = self._attempt_to_build_from_handler(cmd, source_dir_path, env)
+            if process.returncode != 0:
+                raise BuilderError(message=err.decode("utf8").strip())
 
         return out.decode("utf8").strip()
+
+    def _attempt_to_build_from_handler(self, cmd, source_dir_path, env):
+        #Path to the source directory for Go files in a diff directory
+        cmd[-1] = Path(source_dir_path, self.handler)
```
Comment (mndeveci) [inline aws_lambda_builders/workflows/go_modules/builder.py:83]: I am not sure if I understand this correctly to construct the path with `handler` value, can we add an example integration test which will validate this? Happy to help you to drive that.

**Review by mndeveci — CHANGES_REQUESTED** [review]
Thanks for the PR!

I would recommend running `make pr` to find the errors for the CI jobs that is failing now.

**Comment by mildaniel** [comment]
This PR also addresses this issue: https://github.com/aws/aws-sam-cli/issues/2844

File: aws_lambda_builders/workflows/go_modules/workflow.py [hunk 2]
```diff
@@ -32,9 +32,8 @@ def __init__(
 
         output_path = osutils.joinpath(artifacts_dir, handler)
 
-        builder = GoModulesBuilder(
-            osutils, binaries=self.binaries, mode=mode, architecture=self.architecture, trim_go_path=trim_go_path
```
[the remark this answers is not in the document]
Comment (Harleenijjar) [reply aws_lambda_builders/workflows/go_modules/workflow.py:36]: I don't think it was removed but rather the handler was added to the parameter.

[the remark this answers is not in the document]
Comment (Harleenijjar) [reply aws_lambda_builders/workflows/go_modules/workflow.py:36]: That was a mistake. I have added it back. thanks!

[the remark this answers is not in the document]
Comment (Harleenijjar) [reply aws_lambda_builders/workflows/go_modules/workflow.py:36]: Done

File: aws_lambda_builders/workflows/go_modules/builder.py [hunk 3]
```diff
@@ -69,6 +71,17 @@ def build(self, source_dir_path, output_path):
...
         if p.returncode != 0:
-            raise BuilderError(message=err.decode("utf8").strip())
+            LOG.debug("Go files not found. Attempting to build for Go files in a different directory")
+            process, p_out, p_err = self._attempt_to_build_from_handler(cmd, source_dir_path, env)
+            if process.returncode != 0:
+                raise BuilderError(message=err.decode("utf8").strip())
 
         return out.decode("utf8").strip()
+
+    def _attempt_to_build_from_handler(self, cmd, source_dir_path, env):
```
[the remark this answers is not in the document]
Comment (Harleenijjar) [reply aws_lambda_builders/workflows/go_modules/builder.py:83]: Done

File: aws_lambda_builders/workflows/go_modules/builder.py [hunk 4]
```diff
@@ -69,6 +71,17 @@ def build(self, source_dir_path, output_path):
         out, err = p.communicate()
 
         if p.returncode != 0:
-            raise BuilderError(message=err.decode("utf8").strip())
+            LOG.debug("Go files not found. Attempting to build for Go files in a different directory")
```
[the remark this answers is not in the document]
Comment (Harleenijjar) [reply aws_lambda_builders/workflows/go_modules/builder.py:75]: Done

File: aws_lambda_builders/workflows/go_modules/builder.py [hunk 5]
```diff
@@ -69,6 +71,17 @@ def build(self, source_dir_path, output_path):
...
+            process, p_out, p_err = self._attempt_to_build_from_handler(cmd, source_dir_path, env)
+            if process.returncode != 0:
+                raise BuilderError(message=err.decode("utf8").strip())
 
         return out.decode("utf8").strip()
+
+    def _attempt_to_build_from_handler(self, cmd, source_dir_path, env):
+        #Path to the source directory for Go files in a diff directory
+        cmd[-1] = Path(source_dir_path, self.handler)
+        LOG.debug(f"Building for Go files in a different directory")
```
[the remark this answers is not in the document]
Comment (Harleenijjar) [reply aws_lambda_builders/workflows/go_modules/builder.py]: Done

File: aws_lambda_builders/workflows/go_modules/workflow.py [hunk 6]
```diff
@@ -33,8 +33,14 @@ def __init__(
         output_path = osutils.joinpath(artifacts_dir, handler)
 
         builder = GoModulesBuilder(
-            osutils, binaries=self.binaries, mode=mode, architecture=self.architecture, trim_go_path=trim_go_path
+            osutils,
+            binaries=self.binaries,
+            mode=mode,
+            architecture=self.architecture,
+            trim_go_path=trim_go_path,
+            handler=handler,
```
Comment (mndeveci) [inline aws_lambda_builders/workflows/go_modules/workflow.py:41]: nit: just keeping the order in the constructor definition for easy readability
```suggestion
            osutils,
            binaries=self.binaries,
            handler=handler,
            mode=mode,
            architecture=self.architecture,
            trim_go_path=trim_go_path,
```

File: aws_lambda_builders/workflows/go_modules/builder.py [hunk 7]
```diff
@@ -21,7 +22,7 @@ class GoModulesBuilder(object):
 
     LANGUAGE = "go"
 
-    def __init__(self, osutils, binaries, mode=BuildMode.RELEASE, architecture=X86_64, trim_go_path=False):
+    def __init__(self, osutils, binaries, handler, mode=BuildMode.RELEASE, architecture=X86_64, trim_go_path=False):
```
[the remark this answers is not in the document]
Comment (Harleenijjar) [reply aws_lambda_builders/workflows/go_modules/builder.py:25]: ` handler` wasnt present before, that was added.

File: aws_lambda_builders/workflows/go_modules/workflow.py (hunk 6 above)
Comment (Harleenijjar) [reply aws_lambda_builders/workflows/go_modules/workflow.py]: right, done!

**Review by mndeveci — APPROVED** [review]

**Review by mildaniel — APPROVED** [review]
Thanks for the updates!

**Review by mndeveci — APPROVED** [review]

**Pull request merged**